doc/source/ray-observability/user-guides/profiling.md

```diff
@@ -0,0 +1,67 @@
+(profiling)=
+# Profiling
+Profiling is one of the most important debugging tools to diagnose performance, out of memory, hanging, or other application issues.
+Here is a list of common profiling tools you may use when debugging Ray applications. 
+- CPU profiling
+    - py-spy
+- Memory profiling
+    - memray
+- GPU profiling
+    - PyTorch Profiler
+- Ray Task / Actor timeline
+
+If Ray doesn't work with certain profiling tools, try running them without Ray to debug the issues.
+
+(profiling-cpu)=
+## CPU profiling
+CPU profiling for Driver and Worker processes. This helps you understand the CPU usage by different processes and debug unexpectedly high or low usage.
+
+(profiling-pyspy)=
+### py-spy
+[py-spy](https://github.com/benfred/py-spy/tree/master) is a sampling profiler for Python programs. Ray Dashboard has native integration with pyspy:
+
+- It lets you visualize what your Python program is spending time on without restarting the program or modifying the code in any way.
+- It dumps the stacktrace of the running process so that you can see what the process is doing at a certain time. It is useful when programs hangs.
+
+:::{note}
+You may run into permission errors when using py-spy in the docker containers. To fix the issue:
+
+- if you start Ray manually in a Docker container, follow the `py-spy documentation`_ to resolve it. 
+- if you are a KubeRay user, follow the :ref:`guide to configure KubeRay <kuberay-pyspy-integration>` and resolve it.
+:::
+
+Here are the {ref}`steps to use py-spy with Ray and Ray Dashboard <observability-debug-hangs>`.
+
+(profiling-cprofile)=
+### cProfile
+cProfile is Python’s native profiling module to profile the performance of your Ray application.
+
+Here are the {ref}`steps to use cProfile <dashboard-cprofile>`.
+
+(profiling-memory)=
+## Memory profiling
+Memory profiling for Driver and Worker processes. This helps you analyze memory allocations in applications, trace memory leaks, and debug high/low memory or out of memory issues.
```

Comment:
`Memory profiling for Driver and Worker processes.` is not a complete sentence.

Reply:
Edited it. WDYT?